File: android/app/src/main/java/com/mdviewer/app/AndroidBridge.java
package com.mdviewer.app;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.widget.Toast;

public class AndroidBridge {
    private final MainActivity mActivity;
    private final WebView mWebView;
    private final Handler mHandler;
    private final SharedPreferences mPrefs;

    public AndroidBridge(MainActivity activity, WebView webView) {
        this.mActivity = activity;
        this.mWebView = webView;
        this.mHandler = new Handler(Looper.getMainLooper());
        this.mPrefs = activity.getSharedPreferences("mdviewer_prefs", Context.MODE_PRIVATE);
    }

    @JavascriptInterface
    public void openNativeFilePicker() {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                mActivity.openFileChooser();
            }
        });
    }

    @JavascriptInterface
    public void openNativeFolderPicker() {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                mActivity.openFolderChooser();
            }
        });
    }

    @JavascriptInterface
    public void closeFolder() {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                mActivity.closeCurrentFolder();
            }
        });
    }

    @JavascriptInterface
    public String getTreeChildren(String docId) {
        return mActivity.getTreeChildrenJson(docId);
    }

    @JavascriptInterface
    public String readTreeFile(String docId) {
        return mActivity.readTreeFileContent(docId);
    }

    @JavascriptInterface
    public String searchFolder(String query) {
        return mActivity.searchFolderJson(query);
    }

    @JavascriptInterface
    public String readNativeFile(String uriString) {
        return mActivity.readNativeFileContent(uriString);
    }

    @JavascriptInterface
    public void showToast(final String message) {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(mActivity, message, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @JavascriptInterface
    public void copyToClipboard(final String text) {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                ClipboardManager clipboard = (ClipboardManager) mActivity.getSystemService(Context.CLIPBOARD_SERVICE);
                ClipData clip = ClipData.newPlainText("Markdown Content", text);
                if (clipboard != null) {
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(mActivity, "Copied to clipboard", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    @JavascriptInterface
    public void shareText(final String text, final String title) {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                Intent sendIntent = new Intent();
                sendIntent.setAction(Intent.ACTION_SEND);
                sendIntent.putExtra(Intent.EXTRA_TEXT, text);
                sendIntent.setType("text/plain");
                Intent shareIntent = Intent.createChooser(sendIntent, title != null ? title : "Share Markdown");
                mActivity.startActivity(shareIntent);
            }
        });
    }

    @JavascriptInterface
    public void savePreference(String key, String value) {
        mPrefs.edit().putString(key, value).apply();
    }

    @JavascriptInterface
    public String getPreference(String key, String defaultValue) {
        return mPrefs.getString(key, defaultValue);
    }

    @JavascriptInterface
    public void closeApp() {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                mActivity.finish();
            }
        });
    }
}

File: android/app/src/main/java/com/mdviewer/app/AndroidBridge.kt
package com.mdviewer.app

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.widget.Toast

class AndroidBridge(
    private val activity: MainActivity,
    private val webView: WebView
) {
    private val handler = Handler(Looper.getMainLooper())
    private val prefs: SharedPreferences =
        activity.getSharedPreferences("mdviewer_prefs", Context.MODE_PRIVATE)

    @JavascriptInterface
    fun openNativeFilePicker() {
        handler.post { activity.openFileChooser() }
    }

    @JavascriptInterface
    fun openNativeFolderPicker() {
        handler.post { activity.openFolderChooser() }
    }

    @JavascriptInterface
    fun closeFolder() {
        handler.post { activity.closeCurrentFolder() }
    }

    @JavascriptInterface
    fun getRecursiveTree(): String {
        return activity.getRecursiveTreeJson()
    }

    @JavascriptInterface
    fun getTreeChildren(docId: String?): String {
        return activity.getTreeChildrenJson(docId ?: "")
    }

    @JavascriptInterface
    fun readTreeFile(docId: String?): String {
        return activity.readTreeFileContent(docId ?: "")
    }

    @JavascriptInterface
    fun searchFolder(query: String?): String {
        return activity.searchFolderJson(query ?: "")
    }

    @JavascriptInterface
    fun readNativeFile(uriString: String?): String {
        return activity.readNativeFileContent(uriString ?: "")
    }

    @JavascriptInterface
    fun downloadAndInstall(apkUrl: String?, versionName: String?) {
        activity.downloadAndInstallApk(apkUrl ?: "", versionName ?: "")
    }

    @JavascriptInterface
    fun openExternalUrl(url: String?) {
        activity.openWebUrl(url ?: "")
    }

    @JavascriptInterface
    fun checkLatestRelease(): String {
        return activity.checkGitHubRelease()
    }

    @JavascriptInterface
    fun setFullscreen(fullscreen: Boolean) {
        activity.setSystemUiFullscreen(fullscreen)
    }

    @JavascriptInterface
    fun getAppVersion(): String {
        return "1.0.7"
    }

    @JavascriptInterface
    fun showToast(message: String?) {
        if (message.isNullOrEmpty()) return
        handler.post {
            Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
        }
    }

    @JavascriptInterface
    fun copyToClipboard(text: String?) {
        if (text == null) return
        handler.post {
            val clipboard = activity.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
            val clip = ClipData.newPlainText("Markdown Content", text)
            clipboard?.setPrimaryClip(clip)
            Toast.makeText(activity, "Copied to clipboard", Toast.LENGTH_SHORT).show()
        }
    }

    @JavascriptInterface
    fun shareText(text: String?, title: String?) {
        if (text == null) return
        handler.post {
            val sendIntent = Intent(Intent.ACTION_SEND).apply {
                putExtra(Intent.EXTRA_TEXT, text)
                type = "text/plain"
            }
            val shareIntent = Intent.createChooser(sendIntent, title ?: "Share Markdown")
            activity.startActivity(shareIntent)
        }
    }

    @JavascriptInterface
    fun savePreference(key: String?, value: String?) {
        if (key != null) {
            prefs.edit().putString(key, value).apply()
        }
    }

    @JavascriptInterface
    fun getPreference(key: String?, defaultValue: String?): String? {
        return if (key != null) prefs.getString(key, defaultValue) else defaultValue
    }

    @JavascriptInterface
    fun closeApp() {
        handler.post { activity.finish() }
    }
}
